src/transport/Tuple.h

```diff
@@ -0,0 +1,148 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ * @file
+ *
+ * Defines a compound transport type (tuple) that can merge several transports
+ * to look like a single one.
+ */
+#ifndef TRANSPORT_TUPLE_H_
+#define TRANSPORT_TUPLE_H_
+
+#include <tuple>
+#include <type_traits>
+
+#include <transport/Base.h>
+
+namespace chip {
+namespace Transport {
+
+/**
+ * Groups together several transports of different types and presents them as a unified transport.
+ *
+ * The usage intent is to be able to group several distinct transport types and make them look
+ * as a single transport.
+ *
+ * Having an example class definition of:
+ *    Tuple<UDP, UDP, TCP>
+ *
+ * Is equivalent to:
+ *    class TupleOfUdpUdpTcp : public BASE {
+ *       private:
+ *          UDP mUdp1;
+ *          UDP mUdp2;
+ *          TCP mTcp2;
+ *       public:
+ *          Init(UDPListenParameters &, UDPListenParameters&, TCPListenParameters) {...}
+ *          CHIP_ERROR SendMessage(...) override;
+ *          bool CanSendToPeer(...) override;
+ *    }
+ *
+ * The intent of this is to allow applications to use any transport types without CHIP pre-defining
+ * popular mappings (like UDP only, UDP and BLE, BLE only etc.) and without using #ifdefs to create
+ * a single 'standard transport'.
+ */
+template <typename... TransportTypes>
+class Tuple : public Base
+{
+public:
+    CHIP_ERROR SendMessage(const MessageHeader & header, const Transport::PeerAddress & address,
+                           System::PacketBuffer * msgBuf) override
+    {
+        return SendMessageImpl<0>(header, address, msgBuf);
+    }
+
+    bool CanSendToPeer(const Transport::PeerAddress & address) override { return CanSendToPeerImpl<0>(address); }
```

Comment:
Please include method Doxygen documentation.

Style nit: please don't one-line the declaration and implementation.

Reply:
Restyler fights me here.

Reply:
@woody-apple or @rwalker-apple, thoughts or insights here? Do we have a clang-format option that enforces one-liners? If so, can we disable it?

Reply:
it may be a side-effect of line-wrapping, will look

Reply:
isn't achievable with clang-format

Reply:
Yeah, no, I don't think we should change line width either. However, I find it surprising that if @andy31415 makes a method function body in a header in a "traditional" non-one-liner style:

```
rvalue name(params, ...)
{
    return someother_function(params);
}
```
 
that clang-format / restyler forces this to a one-liner.